File: SmartBridge/app/src/main/java/com/captain/smartbridge/UI/Activity/AbsActivity.java
package com.captain.smartbridge.UI.Activity;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;

import com.captain.smartbridge.Common.CommonUtils;
import com.captain.smartbridge.R;

/**
 * Created by Captain on 17/6/28.
 */

public abstract class AbsActivity extends AppCompatActivity {
    MyReceiver receiver = new MyReceiver();
    IntentFilter filter = new IntentFilter();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setSelfContentView();
        prepareDatas();
        initViews();

        regListener();
    }


    protected void regListener() {
        //注册广播接收者
        filter.addAction("exit_app");
        registerReceiver(receiver, filter);
    }

    protected abstract void setSelfContentView();

    protected abstract void prepareDatas();

    protected abstract void initViews();

    protected void readyGo(Class<?> clazz) {
        Intent intent = new Intent(this, clazz);
        startActivity(intent);
    }

    protected void readyGo(Class<?> clazz, Bundle bundle) {
        Intent intent = new Intent(this, clazz);
        if (null != bundle) {
            intent.putExtras(bundle);
        }
        startActivity(intent);
    }


    protected void readyGoThenKill(Class<?> clazz) {
        Intent intent = new Intent(this, clazz);
        startActivity(intent);
        finish();
    }


    protected void readyGoThenKill(Class<?> clazz, Bundle bundle) {
        Intent intent = new Intent(this, clazz);
        if (null != bundle) {
            intent.putExtras(bundle);
        }
        startActivity(intent);
        finish();
    }


    protected void readyGoForResult(Class<?> clazz, int requestCode) {
        Intent intent = new Intent(this, clazz);
        startActivityForResult(intent, requestCode);
    }


    protected void readyGoForResult(Class<?> clazz, int requestCode, Bundle bundle) {
        Intent intent = new Intent(this, clazz);
        if (null != bundle) {
            intent.putExtras(bundle);
        }
        startActivityForResult(intent, requestCode);
    }

    protected void showToast(String msg) {
        if (null != msg && !CommonUtils.isEmpty(msg)) {
            Snackbar.make(getWindow().getDecorView(), msg, Snackbar.LENGTH_SHORT).show();
        }
    }

    protected void showNetWorkError() {
        showToast(getResources().getString(R.string.network_error_tips));
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(receiver);
    }

    class MyReceiver extends BroadcastReceiver{

        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction().equals("exit_app")){
                context.unregisterReceiver(this);
                finish();
            }
        }
    }


}

File: SmartBridge/app/src/main/java/com/captain/smartbridge/UI/Activity/AbsActivity.kt
package com.captain.smartbridge.UI.Activity

import android.content.Intent
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import com.captain.smartbridge.Common.CommonUtils
import com.captain.smartbridge.R

/**
 * Created by fish on 17-4-24.
 */

abstract class AbsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setSelfContentView()
        prepareDatas()
        initViews()
    }

    protected abstract fun setSelfContentView()

    protected abstract fun prepareDatas()

    protected abstract fun initViews()

    protected fun readyGo(clazz: Class<*>) {
        val intent = Intent(this, clazz)
        startActivity(intent)
    }

    protected fun readyGo(clazz: Class<*>, bundle: Bundle?) {
        val intent = Intent(this, clazz)
        if (null != bundle) {
            intent.putExtras(bundle)
        }
        startActivity(intent)
    }

    protected fun readyGoThenKill(clazz: Class<*>) {
        val intent = Intent(this, clazz)
        startActivity(intent)
        finish()
    }

    protected fun readyGoThenKill(clazz: Class<*>, bundle: Bundle?) {
        val intent = Intent(this, clazz)
        if (null != bundle) {
            intent.putExtras(bundle)
        }
        startActivity(intent)
        finish()
    }

    protected fun readyGoForResult(clazz: Class<*>, requestCode: Int) {
        val intent = Intent(this, clazz)
        startActivityForResult(intent, requestCode)
    }

    protected fun readyGoForResult(clazz: Class<*>, requestCode: Int, bundle: Bundle?) {
        val intent = Intent(this, clazz)
        if (null != bundle) {
            intent.putExtras(bundle)
        }
        startActivityForResult(intent, requestCode)
    }

    protected fun showToast(msg: String?) {
        if (null != msg && !CommonUtils.isEmpty(msg)) {
            Snackbar.make(window.decorView, msg, Snackbar.LENGTH_SHORT).show()
        }
    }

    protected fun showNetWorkError() {
        showToast(resources.getString(R.string.network_error_tips))
    }
    //
    //    protected String getInnerErrorInfo(RetrofitError error) {
    //        ErrorRes res = (ErrorRes) error.getBodyAs(ErrorRes.class);
    //        if (res != null) {
    //            return res.getReason();
    //        } else {
    //            return getString(R.string.common_error_msg);
    //        }
    //    }

    //    /**
    //     * toggle show loading
    //     *
    //     * @param toggle
    //     */
    //    protected void toggleShowLoading(boolean toggle, String msg) {
    //        if (null == mVaryViewHelperController) {
    //            throw new IllegalArgumentException("You must return a right target view for loading");
    //        }
    //
    //        if (toggle) {
    //            mVaryViewHelperController.showLoading(msg);
    //        } else {
    //            mVaryViewHelperController.restore();
    //        }
    //    }
    //
    //    /**
    //     * toggle show empty
    //     *
    //     * @param toggle
    //     */
    //    protected void toggleShowEmpty(boolean toggle, String msg, View.OnClickListener onClickListener) {
    //        if (null == mVaryViewHelperController) {
    //            throw new IllegalArgumentException("You must return a right target view for loading");
    //        }
    //
    //        if (toggle) {
    //            mVaryViewHelperController.showEmpty(msg, onClickListener);
    //        } else {
    //            mVaryViewHelperController.restore();
    //        }
    //    }
    //
    //    /**
    //     * toggle show error
    //     *
    //     * @param toggle
    //     */
    //    protected void toggleShowError(boolean toggle, String msg, View.OnClickListener onClickListener) {
    //        if (null == mVaryViewHelperController) {
    //            throw new IllegalArgumentException("You must return a right target view for loading");
    //        }
    //
    //        if (toggle) {
    //            mVaryViewHelperController.showError(msg, onClickListener);
    //        } else {
    //            mVaryViewHelperController.restore();
    //        }
    //    }
    //
    //    /**
    //     * toggle show network error
    //     *
    //     * @param toggle
    //     */
    //    protected void toggleNetworkError(boolean toggle, View.OnClickListener onClickListener) {
    //        if (null == mVaryViewHelperController) {
    //            throw new IllegalArgumentException("You must return a right target view for loading");
    //        }
    //
    //        if (toggle) {
    //            mVaryViewHelperController.showNetworkError(onClickListener);
    //        } else {
    //            mVaryViewHelperController.restore();
    //        }
    //    }

    protected fun filterException(e: Exception?): Boolean {
        if (e != null) {
            e.printStackTrace()
            showToast(e.message)
            return false
        } else {
            return true
        }
    }

    override fun onDestroy() {
        super.onDestroy()
    }
}
